### DISCUSSION TITLE
How do I sync multiple (business) sharepoints?

### DISCUSSION THREAD

**Discussion by patricknoffke** [abraunegg/onedrive#1439] [Q&A] [answered]
Hello,

I have a OneDrive business account, and we have multiple Sharepoint sites (with documents) created within the organization.  These are not shared folders, so they do not show up with "onedrive --list-shared-folders".  I can sync my personal (business) OneDrive account with no issues.  Is there a way I can also sync the documents from other sharepoint sites of which I'm a member?

I have the URL for each site, and I tried adding the last part of it to the business_shared_folders file, and that didn't seem to work.  I can sync each of these to different folders on my Windows or Mac PCs.

Please let me know if you need more information.

Thank you for continuing development on this great tool!

**Comment by abraunegg** [comment, marked as answer]
@patricknoffke 
Please read the documentation on how to sync SharePoint Shared Libraries: https://github.com/abraunegg/onedrive/blob/master/docs/SharePoint-Shared-Libraries.md

**Reply by patricknoffke** [reply]
Thank you, I was able to get both my individual and a shared library to sync!

However, I do not have all my Sharepoint sites, of which I'm a member, listed on my individual sharepoint site.  I don't even have a "Shared Libraries" list shown on my OneDrive page.  All I see is the My Files/Recent/Shared/Recycle Bin, and below that a "Quick Access" list that includes some sites that I have visited recently.

I was able to find one of my sites with the onedrive --get-O365-drive-id

However, for another site (the name has a space and period in it, something like "Mr. Rabbit"), I get this:
```
$ onedrive --verbose --get-O365-drive-id "Mr. Rabbit"
Using 'user' Config Dir: /home/pnoffke/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
No user or system config file found, using application defaults
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/pnoffke/OneDrive
Application version: v2.4.11-5-gb7eedbd
Account Type: business
Default Drive ID: b!vpvxvZig0E6lfxoAv-FYsdq7rlRvQGRJrYuKGpHKz22D_YB90kn2RKlUk9ZQ6ATo
Default Root ID: 0164K6XB56Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 5450623444568
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Office 365 Library Name Query: Mr. Rabbit

ERROR: SharePoint Site details not provided for: myhillrom-my.sharepoint.com
ERROR: The SharePoint Site results returned from OneDrive API do not contain the required items to match. Please check your permissions with your site administrator.
ERROR: Your site security settings is preventing the following details from being accessed: 'displayName' or 'id'
 - Is 'displayName' available = false
 - Is 'id' available          = true
ERROR: To debug this further, please increase verbosity (--verbose or --verbose --verbose) to provide further insight as to what details are actually being returned.
std.json.JSONException@std/json.d(627): Key not found: displayName
----------------
??:? pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x63d473]
??:? pure @safe inout(std.json.JSONValue)* std.exception.enforce!(std.json.JSONException).enforce!(inout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*, lazy const(char)[], immutable(char)[], ulong) [0x6fae11]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(immutable(char)[]) [0x6b13d5]
??:? void sync.SyncEngine.querySiteCollectionForDriveID(immutable(char)[]) [0x6837f9]
??:? _Dmain [0x65b608]
std.net.curl.CurlException@std/net/curl.d(4402): A libcurl function was given a bad argument on handle null
----------------
??:? pure @safe void std.exception.bailOut!(std.net.curl.CurlException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x6fb15e]
??:? pure @safe bool std.exception.enforce!(std.net.curl.CurlException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x6fb0cf]
??:? void std.net.curl.Curl._check(int) [0x6d16e7]
??:? void std.net.curl.Curl.clear(etc.c.curl.CurlOption) [0x6d1a92]
??:? void std.net.curl.HTTP.clearRequestHeaders() [0x6d07c7]
??:? void onedrive.OneDriveApi.shutdown() [0x6613e0]
??:? _Dmain [0x65c920]
```

Is there another way to find the drive ID for a particular site?

Thanks,
Patrick

**Reply by abraunegg** [reply]
@patricknoffke 
The application 'should' not crash like that, so that is a bug that needs to be fixed.

In terms of the site name, you have to search for the exact site name as it is presented online via the WebUI - as this is what the OneDrive API uses as data which can then be searched against.

**Comment by abraunegg** [comment]
@patricknoffke 
In regards to the application crash, please can you test the following PR:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1445/head:pr1445
git checkout pr1445
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be: `onedrive v2.4.11-7-gd091a3a` or greater

The crash is being caused by the SharePoint site not provided the expected data - so this also is an Azure AD permissions issue / SharePoint issue which you *must* fix if you want to search what SharePoint sites are available.

**Reply by patricknoffke** [reply]
I finally was able to test this.  It appears the fix worked and I am able to get the drive ID (I removed the value in the output below).

```
$ ./onedrive --verbose --get-O365-drive-id "Mr. Rabbit"
Using 'user' Config Dir: /home/pnoffke/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
No user or system config file found, using application defaults
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Opening the item database ...
All operations will be performed in: /home/pnoffke/OneDrive
Application version: v2.4.11-7-gd091a3a
Account Type: business
Default Drive ID: b!vpvxvZig0E6lfxoAv-FYsdq7rlRvQGRJrYuKGpHKz22D_YB90kn2RKlUk9ZQ6ATo
Default Root ID: 0164K6XB56Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 1054939674528
Fetching details for OneDrive Root
OneDrive Root exists in the database
Initializing the Synchronization Engine ...
Office 365 Library Name Query: Mr. Rabbit

ERROR: SharePoint Site details not provided for: myhillrom-my.sharepoint.com
ERROR: The SharePoint Site results returned from OneDrive API do not contain the required items to match. Please check your permissions with your site administrator.
ERROR: Your site security settings is preventing the following details from being accessed: 'displayName' or 'id'
 - Is 'displayName' available = false
 - Is 'id' available          = true
ERROR: To debug this further, please increase verbosity (--verbose or --verbose --verbose) to provide further insight as to what details are actually being returned.

-----------------------------------------------
Site Name:    Mr. Rabbit
Library Name: Documents
drive_id:     <removed>
Library URL:  <removed>
-----------------------------------------------

```

Thank you for your help.

**Reply by abraunegg** [reply]
@patricknoffke 
Please do not use the PR version `v2.4.11-7-gd091a3a` as this was merged into master quite a while ago - so you are using an old and unsupported release now ..  rebuild your client from 'master' and use v2.4.13 or greater.

**Reply by patricknoffke** [reply]
I just did so, and that version also works.  Thanks again.

**Comment by wgranicki** [comment]
@abraunegg 
Although this question is marked as answered, I don't see the answer to the main question here. So forgive me for reviving an old thread, but I have a similar issue.

I have several (a dozen or so) SharePoint Libraries shared with me within my organization.
Could you confirm, that the only way to get them all to synchronize is to configure several independent instances of onedrive client (one for each SharePoint Library), and run them concurrently?

If so, do you think it'd make sense to open a Feature Request regarding this issue (I mean: to have one client to sync them all), or is it non-viable?

**Comment by abraunegg** [comment]
@wgranicki 
> I have several (a dozen or so) SharePoint Libraries shared with me within my organization.
Could you confirm, that the only way to get them all to synchronize is to configure several independent instances of onedrive client (one for each SharePoint Library), and run them concurrently?

The only way to support multiple SharePoint Libraries is to configure several independent instances of the client. This is clearly documented here: https://github.com/abraunegg/onedrive/blob/master/docs/SharePoint-Shared-Libraries.md#how-to-configure-multiple-onedrive-sharepoint-shared-library-sync

> If so, do you think it'd make sense to open a Feature Request regarding this issue (I mean: to have one client to sync them all), or is it non-viable?

No - because there are too many challenges to even taking that approach - for instance:
* How the OneDrive API works
* How the local file system is checked for local changes & monitored for changes
* How local files are stored in the database in relation to parent | child relationships that tie back to the account Drive ID
* How the Drive ID -> local folder location then would be mapped either logically, or in code or in some new overlay mapping to logically untangle this before processing
* How the API calls are different for root objects for lookup
* How the API calls are different to determine an objects location on OneDrive based on the DriveID
* How the API calls are different to determine an objects location on OneDrive based on the local path .. which now would be different to its actual location

The only path is for you to have separate configurations for each SharePoint Library. Sorry if this is inconvenient for you.
